lib/provisioning/Provisioner.js

```diff
@@ -129,24 +143,170 @@ Provisioner.prototype._validate = function(actual, parameterName) {
+            'key'
+        ];
+    }
     for (var i = 0; i < parameterNames.length; i++) {
         this._validate(parameters[parameterNames[i]], parameterNames[i]);
     }
 };
 
+Provisioner.prototype._updateBridgingState = Promise.coroutine(
+    function*(roomId, bridger, status, skey) {
+        let intent = this._ircBridge.getAppServiceBridge().getIntent();
+        try {
+            yield intent.client.sendStateEvent(roomId, 'm.room.bridging', {
+                bridger: bridger,
+                status: status // pending, success, failure
+            }, skey);
+        }
+        catch (err) {
+            throw new Error(`Could not update m.room.bridging state in this room ($err.message)`);
+        }
+    }
+);
+
+Provisioner.prototype._authoriseProvisioning = Promise.coroutine(
+    function*(server, bridger, ircChannel, roomId, opNick, key) {
+        let ircDomain = server.domain;
+
+        // (Matrix) Check power level of user
+        let matrixClient = this._ircBridge.getAppServiceBridge().getClientFactory().getClientAs();
+        var lookup = new StateLookup({
+            client : matrixClient,
+            eventTypes: ['m.room.power_levels']
+        });
+
+        yield lookup.trackRoom(roomId);
+
+        let powerState = lookup.getState(
+            roomId,
+            'm.room.power_levels'
+        )[0];
+
+        let actualPower = powerState.content.users[bridger];
+        let requiredPower = powerState.content.events["m.room.power_levels"];
+
+        if (actualPower < requiredPower) {
+            throw new Error('User does not possess high enough power level');
+        }
+
+        // (IRC) Check that op's nick is actually op
+        let info = yield this._ircBridge.checkNickExists(server, opNick);
+
+        if (!info.isOp) {
+            throw new Error('Provided user is not an Op');
+        }
+
+        // (Matrix) update room state
+        // State key for m.room.bridging
+        let skey = `${ircDomain}${ircChannel}`;
+
+        // Send pending m.room.bridging
+        this._updateBridgingState(roomId, bridger, 'pending', skey);
+
+        // (IRC) Ask operator for authorisation
+        // Time that operator has to respond before giving up
+        let timeout = 10;
+
+        // Send PM to Bob
+        let botClient = yield this._ircBridge.getBotClient(server);
+        let authRequestAction = new IrcAction("message",
+            `${bridger} has requested to bridge ${roomId} with ${ircChannel} on this IRC ` +
+            `network. Respond with 'yes' to allow, or simply ignore this message to ` +
+            `dissallow. You have ${timeout} seconds from when this message was sent.`);
+        let pmRoom = yield botClient.joinChannel(opNick, key);
+
+        // Promise to receive PM from Bob or timeout
+        let opReplyPromise = this.expectPm(server, opNick, timeout);
+
+        this._ircBridge.sendIrcAction(pmRoom, botClient, authRequestAction);
+
+        opReplyPromise.then(
```

Comment:
But you're in a coroutine already :( :( :( just say no to `then` and accept `yield` into your heart.

```javascript
try {
  yield opReplyPromise;
  // fulfilled function code here
}
catch (err) {
  // rejected function code here
}
```